cpp/src/request_tag.cpp

```diff
@@ -56,11 +56,13 @@ RequestTag::RequestTag(std::shared_ptr<Component> endpointOrWorker,
               transferDirection,
               buffer,
               length,
-              DelayedSubmissionData(DelayedSubmissionOperationType::Tag,
-                                    transferDirection,
-                                    transferDirection == TransferDirection::Send
-                                      ? DelayedSubmissionTag(tag, std::nullopt)
-                                      : DelayedSubmissionTag(tag, tagMask))),
+              transferDirection == TransferDirection::Send
+                ? DelayedSubmissionData(DelayedSubmissionOperationType::Tag,
+                                        transferDirection,
+                                        DelayedSubmissionTagSend(buffer, length, tag))
+                : DelayedSubmissionData(DelayedSubmissionOperationType::Tag,
+                                        transferDirection,
+                                        DelayedSubmissionTagReceive(buffer, length, tag, tagMask))),
```

Comment:
There _should_ be no need for the `DelayedSubmissionOperationType` and `transferDirection` arguments any more, because that information is encoded in the particular type in the `variant`, so this is sending in redundant information which may not match.